addon/-private/decorators/default-if-undefined.js

```diff
@@ -0,0 +1,36 @@
+import { decorator } from '@ember-decorators/utils/decorator';
+
+const VALUES = new WeakMap();
+
+function getValues(object) {
+  if (!VALUES.has(object)) {
+    VALUES.set(object, Object.create(null));
+  }
+
+  return VALUES.get(object);
+}
+
+export default decorator(({ initializer, key, descriptor, ...desc }) => {
+  delete descriptor.writable;
+  return {
+    ...desc,
+    key,
+    kind: 'method',
+    descriptor: {
+      ...descriptor,
+      get() {
+        const values = getValues(this);
+
+        if (!(key in values)) {
+          values[key] = initializer.call(this);
```

Comment:
This has the following implications:

- the initializer is invoked lazily on access, not eagerly on class instantiation
- the initializer is called repeatedly, whenever a default value is needed, meaning it will not be cached

Suppose:

```
class Foo {
  @defaultIfUndefined bar = Math.random();
}
```

Whenever `bar` gets set to `undefined`, it will receive a new random value.

The actual `@defaultIfUndefined` decorator will likely differ from this or allow this to be configurable. But for your use case, this implementation should be sufficient. I decided against caching and eager evaluation, because it would further complicate things, because some nice features are still missing from the spec.